GRADLE.md

```diff
@@ -0,0 +1,88 @@
+The most visible task is the `server-jetty-app:run` task, but other tasks like code generation use the runtime toolchain
+as well.
+
+* `deephaven.runtimeVersion`: the runtime Java version to use, by default is 11
+* `deephaven.runtimeVendor`: the runtime Java vendor to use
+
+For example, the following would ensure that the server is run with JDK 21 from a JVM vendor name that contains "azul":
+
+`./gradlew server-jetty-app:run -Pdeephaven.runtimeVersion=21 -Pdeephaven.runtimeVendor=azul`
+
+### Test Runtime Toolchain
+
+The test runtime toolchains are used to invoke "test" java tasks; that is, tasks of `org.gradle.api.tasks.testing.Test`.
+
+* `deephaven.testRuntimeVersion`: the test runtime Java version to use, by default is 11
+* `deephaven.testRuntimeVendor`: the test runtime Java vendor to use
+
+The [nightly checks](.github/workflows/nightly-check-ci.yml) takes advantage of this to test multiple Java version for
+unit testing.
+
+For example, the following will run the test tasks with JDK 21 from a JVM vendor name that contains "amazon":
+
+`./gradlew test -Pdeephaven.testRuntimeVersion=21 -Pdeephaven.runtimeVendor=amazon`
+
+### Compiler Toolchain
+
+The compiler toolchains are used to invoke "compile" java tasks; that is, tasks of `org.gradle.api.tasks.compile.JavaCompile`.
+
+* `deephaven.compilerVersion`: the compile Java version to use, by default is 11
+* `deephaven.compilerVendor`: the compile Java vendor to use
+
+Note: these gradle properties control both compiling of the main source sets and test source sets.
+
+### Language Levels
+
+A language level is the Java language level the source is written for. This dictates the minimum version required to run
+the specific project. Depending on context, some subprojects have a different language level than other subprojects. For
+example, `server-jetty-app` has a language level of 11 - it must be run with Java 11 or greater;
+`java-client-session` has a language level of 8 - it must be run with Java 8 or greater.
+
+* `deephaven.languageLevel`: the language level for the main source set
+* `deephaven.testLanguageLevel`: the language level for the test source set
+
+The language levels are mostly internal properties of the individual subprojects.
+
+### Application options
+
+* `deephaven.javaOpts`: the "generally applicable and recommended JVM options" for the application. Currently, defaults
+to `-XX:+UseG1GC -XX:MaxGCPauseMillis=100 -XX:+UseStringDeduplication`.
+
+This property is mainly meant to serve as the defaults for `org.gradle.api.plugins.JavaApplication` tasks, which in
+turn serves as the defaults for users running the application (it also applies to `org.gradle.api.tasks.JavaExec`
+and `org.gradle.api.tasks.testing.Test` tasks). Overly specific options do not belong here. For example, heap settings
+(`-Xmx4g`) or system properties (`-Dkey=value`) should not be set here.
```

Comment:
How are these "overly specific" settings set?

Reply:
Typically, we expect that end users will specify their heap size explicitly when necessary, and otherwise will inherit some base percentage of RAM. For example, Deephaven can be used in low-memory systems (I've done as low as 512MB), but if we set any specific amount, that becomes an artificial floor. A bit more dramatic an example, what about `-Xmx256g`? That seems arbitrarily large, just as `-Xmx4g` seems arbitrary.

If we are tempted to set a system property on all deployments by default, we should prefer to set that at the `dh-defaults.prop` level as opposed to a command line argument on application startup.

Reply:
I read the text differently.  I interpreted the text block to mean that some types of settings couldn't be set there and another path was needed to set them, so my question was how to set them.  Maybe reread this and make sure the intention is clear in the text.